ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java

```diff
@@ -64,64 +63,70 @@ public MainnetBlockValidator(
    *     the block, empty if the block was deemed invalid or couldn't be processed
    */
   @Override
-  public Optional<BlockProcessingOutputs> validateAndProcessBlock(
+  public BlockValidator.Result validateAndProcessBlock(
       final ProtocolContext context,
       final Block block,
       final HeaderValidationMode headerValidationMode,
       final HeaderValidationMode ommerValidationMode) {
+
     final BlockHeader header = block.getHeader();
 
     final MutableBlockchain blockchain = context.getBlockchain();
     final Optional<BlockHeader> maybeParentHeader =
         blockchain.getBlockHeader(header.getParentHash());
     if (maybeParentHeader.isEmpty()) {
-      LOG.error(
-          "Attempted to import block {} with hash {} but parent block {} was not present",
-          header.getNumber(),
-          header.getHash(),
-          header.getParentHash());
+      String errMsg =
+          "Attempted to import block "
+              + block.toLogString()
+              + ", but parent block "
+              + header.getParentHash()
+              + " was not present";
+      LOG.error(errMsg);
       badBlockManager.addBadBlock(block);
-      return Optional.empty();
+      return new Result(errMsg);
     }
     final BlockHeader parentHeader = maybeParentHeader.get();
 
     if (!blockHeaderValidator.validateHeader(header, parentHeader, context, headerValidationMode)) {
       badBlockManager.addBadBlock(block);
-      return Optional.empty();
+      return new Result("Invalid header for block " + block.toLogString());
```

Comment:
@garyschulte is it enough this generic `Invalid header for block` message or we should popup more details from `blockHeaderValidator.validateHeader`?

Reply:
:+1:

Reply:
Ok, then we can stop here for the moment, I have just made un update to extract the failure handling code in a method, so the code is cleaner.